### PR TITLE
feat: add preflight checks for macos

### LINKED ISSUE #410: Provide pre-flight checks for Mac
### Is your feature request related to a problem? Please describe

Similarly to the work done for [windows](https://github.com/containers/podman-desktop/issues/326), we need to provide pre-flight checks for installation of Podman on Mac. 

### Describe the solution you'd like

We should check for:
- Version of the OS (if supported)
- Is a previous version already installed? and potentially already running?
- Are there enough resources? 

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by evidolob**
### What does this PR do?
Adds preflight check for podman install on macOS.

### Screenshot/screencast of this PR
Screenshot 2022-09-01 at 16 18 13

### What issues does this PR fix or reference?
Resolve: https://github.com/containers/podman-desktop/issues/410

### CODE REVIEW

File: extensions/podman/src/macos-checks.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import * as os from 'node:os';
+import * as compareVersions from 'compare-versions';
+
+export class MacCPUCheck extends BaseCheck {
+  title = 'CPU';
+  private readonly MIN_CPU_NUMBER = 2;
+  async execute(): Promise<extensionApi.CheckResult> {
+    const cpus = os.cpus();
+    if (cpus.length < this.MIN_CPU_NUMBER) {
+      return this.createFailureResult(`You need to have at least ${this.MIN_CPU_NUMBER} CPU cores to run podman.`);
```
Comment (benoitf) [inline extensions/podman/src/macos-checks.ts]:
```suggestion
      return this.createFailureResult(`You need to have at least ${this.MIN_CPU_NUMBER} CPU cores to install podman.`);
```

suggestion, not sure what is the better

File: extensions/podman/src/macos-checks.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+export class MacMemoryCheck extends BaseCheck {
+  title = 'RAM';
+  private readonly REQUIRED_MEM = 4 * 1024 * 1024 * 1024; // 4Gb
+
+  async execute(): Promise<extensionApi.CheckResult> {
+    const totalMem = os.totalmem();
+    if (this.REQUIRED_MEM <= totalMem) {
+      return this.createSuccessfulResult();
+    } else {
+      return this.createFailureResult('You need at least 4GB to run Podman.');
```
Comment (benoitf) [inline extensions/podman/src/macos-checks.ts]:
```suggestion
      return this.createFailureResult('You need at least 4GB to install Podman.');
```

File: extensions/podman/src/macos-checks.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  title = 'macOS Version';
+  private readonly MINIMUM_VERSION = '19.0.0'; // first macOS Catalina kernel version
+
+  async execute(): Promise<extensionApi.CheckResult> {
+    const darwinVersion = os.release();
+    if (compareVersions.compare(darwinVersion, this.MINIMUM_VERSION, '>=')) {
+      return this.createSuccessfulResult();
+    }
+
+    return this.createFailureResult('To be able to run podman you need to update to macOS Catalina.');
```
Comment (benoitf) [inline extensions/podman/src/macos-checks.ts]:
```suggestion
    return this.createFailureResult('To be able to install podman you need to update to macOS Catalina.');
```

**Review by benoitf — COMMENTED** [review]
I've not yet tested

**Comment by benoitf** [comment]
@evidolob I do not see the check for 'updating' when Podman has been previously installed using brew.
I think we shouldn't proceed in that case and notify the user to use brew to perform the update or uninstall previously the brew version

**Comment by evidolob** [comment]
@benoitf Now Update has own preflight checks:
Screenshot 2022-09-05 at 12 19 22
and if user has podman installed with brew we ask to use brew to update podman.

File: extensions/podman/src/macos-checks.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  title = 'Podman Installation';
+  async execute(): Promise<extensionApi.CheckResult> {
+    const ownPodmanInstalled = await this.checkFileExists('/opt/podman/bin/podman');
+    if (!ownPodmanInstalled) {
+      return this.createFailureResult(
+        'You have podman installed with "brew", run "brew update && brew upgrade podman" to install new version',
+        'Brew Documentation',
+        'https://docs.brew.sh/Manpage#upgrade-options-outdated_formulaoutdated_cask-',
+      );
+    }
```
Comment (benoitf) [inline extensions/podman/src/macos-checks.ts]: I would say that we can't assume podman has been installed using brew if we don't find podman in `/opt/podman/bin` folder as people may have install it differently

`brew list --verbose podman` will return 0 if installed or 1 if not installed so it's look like a better check to know if podman is being installed using brew or not

**Review by benoitf — APPROVED** [review]
Checked on macOS and working as expected

But please rework the brew check by using brew commands instead of assuming that if it's not in the /opt directory it's using brew

File: extensions/podman/src/macos-checks.ts (hunk 4 above)
Comment (benoitf) [reply extensions/podman/src/macos-checks.ts]: `HOMEBREW_NO_AUTO_UPDATE=1` and `HOMEBREW_NO_ANALYTICS=1` should be used to get faster result

**Comment by benoitf** [comment]
Hi,
I think it's still missing https://github.com/containers/podman-desktop/blob/main/extensions/podman/src/podman-cli.ts#L55-L57

https://user-images.githubusercontent.com/436777/188847607-ba598096-aa8e-4ec6-893b-28357aed7f8c.mp4

**Review by benoitf — APPROVED** [review]
image

working as expected 👍

**Pull request merged**